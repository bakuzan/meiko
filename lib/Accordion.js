import classNames from 'classnames';
import React, { useState, useEffect } from 'react';

import { usePrevious } from './hooks/usePrevious';
import styles from './styles/Accordion';

function AccordionToggle({ className, heading, onToggle }) {
  return (
    <button type="button" className={className} onClick={onToggle}>
      {heading}
    </button>
  );
}

function Accordion({
  children,
  className,
  contentProps = {},
  defaultIsCollapsed,
  headingProps = {},
  heading,
  toggleComponent = AccordionToggle,
  onToggle,
  ...pass
}) {
  const [isCollapsed, setCollapsed] = useState(defaultIsCollapsed ?? true);
  const prevCollapsed = usePrevious(isCollapsed);

  const Toggler = toggleComponent;

  useEffect(() => {
    const hasPrev = prevCollapsed !== undefined;
    const hasNew = prevCollapsed !== isCollapsed;

    if (onToggle && hasPrev && hasNew) {
      onToggle(isCollapsed);
    }
  }, [onToggle, isCollapsed, prevCollapsed]);

  return (
    <div
      {...pass}
      className={classNames('accordion', styles.accordion, className)}
    >
      <div
        {...headingProps}
        className={classNames(
          'accordion__heading',
          styles.accordion__heading,
          headingProps.className
        )}
      >
        <Toggler
          className={classNames(
            'accordion__toggle',
            styles.accordion__toggle,
            isCollapsed && [
              'accordion__toggle--checked',
              styles.accordion__toggle_collapsed
            ]
          )}
          heading={heading}
          onToggle={() => setCollapsed((p) => !p)}
        />
      </div>
      <div
        {...contentProps}
        aria-hidden={isCollapsed}
        className={classNames(
          'accordion__content',
          styles.accordion__content,
          contentProps.className,
          isCollapsed && [
            'accordion__content--collapsed',
            styles.accordion__content_collapsed
          ]
        )}
      >
        {typeof children !== 'function' ? children : children(isCollapsed)}
      </div>
    </div>
  );
}

Accordion.propTypes = {
  heading: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.node
  ]).isRequired,
  defaultIsCollapsed: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.node
  ]),
  onToggle: PropTypes.func
};

export default Accordion;
