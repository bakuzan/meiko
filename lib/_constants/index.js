export { default as Strings } from './strings';
export { default as Types } from './types';
export { default as Urls } from './urls';
export { default as Icons } from './icons';
export { PositionEnum, KeyCodes, OPEN_KEYS, CLOSE_KEYS } from './enums';
