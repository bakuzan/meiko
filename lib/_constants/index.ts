import _Strings from './strings';
import _Types from './types';
import _Urls from './urls';
import _Icons from './icons';
import * as _Enums from './enums';

export const Strings = _Strings;
export const Types = _Types;
export const Urls = _Urls;
export const Icons = _Icons;
export const Enums = _Enums;
