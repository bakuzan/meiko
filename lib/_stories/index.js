import React from 'react';
// import { storiesOf } from '@storybook/react';

import './styles/index.scss';

// storiesOf('Index', module).add('basic', () => <div>Exists to load styles</div>);

export const storyStyle = {
  maxWidth: '300px',
  margin: 'auto',
  marginTop: '20px'
};

const styles = { display: 'flex', justifyContent: 'center', marginTop: '30px' };
export const CenterDecorator = (storyFn) => (
  <div style={styles}>{storyFn()}</div>
);
