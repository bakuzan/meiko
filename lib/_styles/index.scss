@import './partials/variables.scss';
@import './box-model.scss';
@import './list.scss';
@import './input-controls.scss';

.button-group {
  display: flex;
  justify-content: space-around;
  align-items: center;
  padding: 5px;
  margin: 5px 0;

  &.centered {
    justify-content: center;
  }
  &.right-aligned {
    justify-content: flex-end;
  }
}
