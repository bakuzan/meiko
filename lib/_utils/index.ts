import adjustDateMonth from 'ayaka/adjustDateMonth';
import adjustDateYear from 'ayaka/adjustDateYear';
import { capitalise } from 'ayaka/capitalise';
import checkIfDatePartsMatch from 'ayaka/checkIfDatePartsMatch';
import convertToBase64 from 'ayaka/convertToBase64';
import createListeners from 'ayaka/createListeners';
import debounce from 'ayaka/debounce';
import formatDateForInput from 'ayaka/formatDateForInput';
import generateUniqueId from 'ayaka/generateUniqueId';
import getDaysInMonthForDate from 'ayaka/getDaysInMonthForDate';
import getDifferenceFromMonday from 'ayaka/daysDifferenceFromMonday';
import getDifferenceFromSunday from 'ayaka/daysDifferenceFromSunday';
import getElementCoordinates from 'ayaka/getElementCoordinates';
import getEventValue from 'ayaka/getEventValue';
import getFirstDateOfMonth from 'ayaka/getFirstDateOfMonth';
import getLastDateOfMonth from 'ayaka/getLastDateOfMonth';
import getMonthName from 'ayaka/getMonthName';
import { getTimeoutSeconds } from 'ayaka/getTimeout';
import isArray from 'ayaka/isArray';
import { isBefore, isBeforeOrEqual } from 'ayaka/isBefore';
import isNumber from 'ayaka/isNumber';
import isObject from 'ayaka/isObject';
import isString from 'ayaka/isString';
import { lighten, darken } from 'ayaka/palette';

export {
  adjustDateMonth,
  adjustDateYear,
  capitalise,
  convertToBase64,
  createListeners,
  debounce,
  generateUniqueId,
  getElementCoordinates,
  getEventValue,
  getTimeoutSeconds,
  isArray,
  isNumber,
  isObject,
  isString,
  checkIfDatePartsMatch,
  formatDateForInput,
  getMonthName,
  getDaysInMonthForDate,
  getFirstDateOfMonth,
  getLastDateOfMonth,
  getDifferenceFromMonday,
  getDifferenceFromSunday,
  isBefore,
  isBeforeOrEqual,
  lighten,
  darken
};
