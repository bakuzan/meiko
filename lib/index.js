// Components
export { default as Alert } from './Alert';
export { default as AppInformation } from './AppInformation';
export { default as AutocompleteInput } from './AutocompleteInput';
export { Button, withButtonisation } from './Button';
export { default as ChipListInput } from './ChipListInput';
export { default as ClearableInput } from './ClearableInput';
export { default as DateSelector } from './DateSelector';
export { default as Dialog } from './Dialog';
export { default as DropdownMenu } from './DropdownMenu';
export { default as FileUploader } from './FileUploader';
export { default as Form } from './Form';
export { default as FormControls } from './FormControls';
export { default as Grid } from './Grid';
export { default as Header } from './Header';
export { default as Image } from './Image';
export { default as ImageSelector } from './ImageSelector';
export { default as List } from './List';
export { SimpleLoading, Loading } from './Loadable';
export { default as LoadableContent } from './LoadableContent';
export { default as LoadingBouncer } from './LoadingBouncer';
export { default as LoadingSpinner } from './LoadingSpinner';
export { default as Logo } from './Logo';

export { default as MultiSelect } from './MultiSelect';
export { default as RadioButton } from './RadioButton';
export { default as RadioToggle } from './RadioToggle';
export { default as RatingControl } from './RatingControl';
export { default as RequestIndicator } from './RequestIndicator';
export { default as ScrollTopButton } from './ScrollTopButton';
export { default as SelectBox } from './SelectBox';
export { default as Sidebar } from './Sidebar';
export { default as Tabs } from './Tabs';
export { default as TagCloudSelector } from './TagCloudSelector';
export { default as Tickbox } from './Tickbox';
export { default as Toaster } from './Toaster';
export { default as Tooltip } from './Tooltip';

export { default as NewTabLink } from './NewTabLink';
export { default as Portal } from './Portal';

// Hooks
export { useGlobalStyles } from './hooks/useGlobalStyles';

// Utils
export { default as toaster, ToasterService } from './utils/toasterService';
export { default as addOutsideClick } from './utils/addOutsideClick';

export {
  PositionEnum,
  KeyCodes,
  OPEN_KEYS,
  CLOSE_KEYS,
  Icons,
  Strings,
  Types,
  Urls
} from './constants';
