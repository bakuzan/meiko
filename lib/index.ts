import './styles/index.scss';

import * as Components from './components';
import { default as fetchFromServer, handleErrorResponse } from './utils/fetch';
import * as Constants from './constants';

export default {
  ...Components,
  Constants,
  fetchFromServer,
  handleErrorResponse
};
