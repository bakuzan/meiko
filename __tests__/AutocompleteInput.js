import { fireEvent } from '@testing-library/react';
import React from 'react';

import { AutocompleteInput } from '../lib';
import textMatcher from './__helpers__/getByTextComplex';

const attr = 'text';
const filter = '';
const items = [
  { id: 5, text: 'Bat' },
  { id: 6, text: 'Cat' },
  { id: 9, text: 'Frog' }
];
const mockChangeFn = jest.fn();
const mockSelectFn = jest.fn();

afterEach(() => jest.resetAllMocks());

it('should render with minimum props', function () {
  const { container } = render(
    <AutocompleteInput
      id="jest"
      attr={attr}
      filter={filter}
      items={items}
      onChange={mockChangeFn}
      onSelect={mockSelectFn}
    />
  );

  expect(container).toMatchSnapshot();
});

it('should render filtered items when filter updated', async function () {
  const { container, getByText, getByTitle, queryByText, rerender } = render(
    <AutocompleteInput
      id="jest"
      attr={attr}
      filter={filter}
      items={items}
      onChange={mockChangeFn}
      onSelect={mockSelectFn}
      clearableInputProps={{ title: 'testInput' }}
    />
  );

  expect(container).toMatchSnapshot();

  // Check the display changes with the filter update
  rerender(
    <AutocompleteInput
      id="jest"
      attr={attr}
      filter={'at'}
      items={items}
      onChange={mockChangeFn}
      onSelect={mockSelectFn}
      clearableInputProps={{ title: 'testInput' }}
    />
  );

  const user = userEvent.setup();
  await user.click(getByTitle('testInput'));

  expect(getByText(textMatcher('Bat'))).toBeTruthy();
  expect(getByText(textMatcher('Cat'))).toBeTruthy();
  expect(queryByText(textMatcher('Frog'))).toBeNull();
  expect(container).toMatchSnapshot();
});

it('should call onChange when filter updated', async function () {
  const { getByTitle } = render(
    <AutocompleteInput
      id="jest"
      attr={attr}
      filter={filter}
      items={items}
      onChange={mockChangeFn}
      onSelect={mockSelectFn}
      clearableInputProps={{ title: 'testInput' }}
    />
  );

  fireEvent.change(getByTitle('testInput'), { target: { value: 'at' } });

  expect(mockChangeFn).toHaveBeenCalled();
});

it('should call onSelect when suggestion is click', async function () {
  const { container, getByText, getByTitle } = render(
    <AutocompleteInput
      id="jest"
      attr={attr}
      filter={'at'}
      items={items}
      onChange={mockChangeFn}
      onSelect={mockSelectFn}
      clearableInputProps={{ title: 'testInput' }}
    />
  );

  const user = userEvent.setup();
  await user.click(getByTitle('testInput'));
  await user.click(getByText(textMatcher('Bat')));

  expect(mockSelectFn).toHaveBeenCalledWith(5);
  expect(container).toMatchSnapshot();
});
