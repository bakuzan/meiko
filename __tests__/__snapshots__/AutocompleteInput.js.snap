// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call onChange when filter updated 1`] = `
<AutocompleteInput
  attr="text"
  clearableInputProps={Object {}}
  disableLocalFilter={false}
  filter=""
  id="jest"
  items={
    Array [
      Object {
        "id": 5,
        "text": "Bat",
      },
      Object {
        "id": 6,
        "text": "Cat",
      },
      Object {
        "id": 9,
        "text": "Frog",
      },
    ]
  }
  noSuggestionsItem={null}
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "target": Object {
              "value": "at",
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onSelect={[MockFunction]}
  suggestionTemplate={[Function]}
>
  <div
    className="autocomplete  mko_ydt0iq-autocomplete"
  >
    <ClearableInput
      id="jest"
      maxNumberText={[Function]}
      name="text"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      value=""
    >
      <div
        className="clearable-input has-float-label  mko_1wcuo11-clearableInput  mko_1aosyyv-controlContainer"
      >
        <div
          className="clearable-input__inner  mko_1wcuo11-clearableInput__inner"
        >
          <ForwardRef(Input)
            className="clearable-input__input  mko_1wcuo11-clearableInput__input"
            id="jest"
            name="text"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            value=""
          >
            <input
              autoComplete="off"
              className="input clearable-input__input  mko_1wcuo11-clearableInput__input"
              id="jest"
              name="text"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder=" "
              type="text"
              value=""
            />
          </ForwardRef(Input)>
          <label
            htmlFor="jest"
          />
        </div>
      </div>
    </ClearableInput>
  </div>
</AutocompleteInput>
`;

exports[`should call onSelect when suggestion is click 1`] = `
<AutocompleteInput
  attr="text"
  clearableInputProps={Object {}}
  disableLocalFilter={false}
  filter="at"
  id="jest"
  items={
    Array [
      Object {
        "id": 5,
        "text": "Bat",
      },
      Object {
        "id": 6,
        "text": "Cat",
      },
      Object {
        "id": 9,
        "text": "Frog",
      },
    ]
  }
  noSuggestionsItem={null}
  onChange={[MockFunction]}
  onSelect={
    [MockFunction] {
      "calls": Array [
        Array [
          5,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  suggestionTemplate={[Function]}
>
  <div
    className="autocomplete  mko_ydt0iq-autocomplete"
  >
    <ClearableInput
      id="jest"
      maxNumberText={[Function]}
      name="text"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      value="at"
    >
      <div
        className="clearable-input has-float-label  mko_1wcuo11-clearableInput  mko_1aosyyv-controlContainer"
      >
        <div
          className="clearable-input__inner  mko_1wcuo11-clearableInput__inner"
        >
          <ForwardRef(Input)
            className="clearable-input__input  mko_1wcuo11-clearableInput__input"
            id="jest"
            name="text"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            value="at"
          >
            <input
              autoComplete="off"
              className="input clearable-input__input  mko_1wcuo11-clearableInput__input"
              id="jest"
              name="text"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder=" "
              type="text"
              value="at"
            />
          </ForwardRef(Input)>
          <label
            htmlFor="jest"
          />
          <Button
            aria-label="Clear input"
            btnSize="small"
            className="clearable-input__clear  mko_1wcuo11-clearableInput__clear"
            icon="❌︎"
            onClick={[Function]}
            type="button"
          >
            <button
              aria-label="Clear input"
              className="button ripple button--icon button--small  mko_xh0n1r-button  mko_xh0n1r-button_icon clearable-input__clear  mko_1wcuo11-clearableInput__clear"
              icon="❌︎"
              onClick={[Function]}
              type="button"
            />
          </Button>
        </div>
      </div>
    </ClearableInput>
    <ul
      className="autocomplete__menu  mko_ydt0iq-autocomplete__menu"
    >
      <AutocompleteSuggestionItem
        activeSuggestion={0}
        attr="text"
        highlightMatch={[Function]}
        index={0}
        item={
          Object {
            "id": 5,
            "text": "Bat",
          }
        }
        key="5"
        selectAutocompleteSuggestion={[Function]}
      >
        <li
          className="autocomplete-suggestion autocomplete-suggestion--active  mko_ydt0iq-suggestion"
        >
          <Button
            className="autocomplete-suggestion__button"
            onClick={[Function]}
            title="Bat"
            type="button"
          >
            <button
              className="button ripple button--standard  mko_xh0n1r-button  mko_xh0n1r-button_standard autocomplete-suggestion__button"
              onClick={[Function]}
              title="Bat"
              type="button"
            >
              <span
                className="autocomplete-suggestion__text  mko_ydt0iq-suggestion__text"
              >
                B
                <span
                  className="autocomplete-suggestion-highlight  mko_ydt0iq-suggestionHighlight"
                >
                  at
                </span>
              </span>
            </button>
          </Button>
        </li>
      </AutocompleteSuggestionItem>
      <AutocompleteSuggestionItem
        activeSuggestion={0}
        attr="text"
        highlightMatch={[Function]}
        index={1}
        item={
          Object {
            "id": 6,
            "text": "Cat",
          }
        }
        key="6"
        selectAutocompleteSuggestion={[Function]}
      >
        <li
          className="autocomplete-suggestion  mko_ydt0iq-suggestion"
        >
          <Button
            className="autocomplete-suggestion__button"
            onClick={[Function]}
            title="Cat"
            type="button"
          >
            <button
              className="button ripple button--standard  mko_xh0n1r-button  mko_xh0n1r-button_standard autocomplete-suggestion__button"
              onClick={[Function]}
              title="Cat"
              type="button"
            >
              <span
                className="autocomplete-suggestion__text  mko_ydt0iq-suggestion__text"
              >
                C
                <span
                  className="autocomplete-suggestion-highlight  mko_ydt0iq-suggestionHighlight"
                >
                  at
                </span>
              </span>
            </button>
          </Button>
        </li>
      </AutocompleteSuggestionItem>
    </ul>
  </div>
</AutocompleteInput>
`;

exports[`should render with minimum props 1`] = `
<div
  className="autocomplete  mko_ydt0iq-autocomplete"
>
  <ClearableInput
    id="jest"
    maxNumberText={[Function]}
    name="text"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    value=""
  />
</div>
`;
